src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -1190,11 +1193,86 @@ class ShenandoahEvacuationTask : public WorkerTask {
   }
 };
 
+class ShenandoahRetireGCLABClosure : public ThreadClosure {
+private:
+  bool const _resize;
+public:
+  explicit ShenandoahRetireGCLABClosure(bool resize) : _resize(resize) {}
+  void do_thread(Thread* thread) override {
+    PLAB* gclab = ShenandoahThreadLocalData::gclab(thread);
+    assert(gclab != nullptr, "GCLAB should be initialized for %s", thread->name());
+    gclab->retire();
+    if (_resize && ShenandoahThreadLocalData::gclab_size(thread) > 0) {
+      ShenandoahThreadLocalData::set_gclab_size(thread, 0);
+    }
+
+    if (ShenandoahHeap::heap()->mode()->is_generational()) {
+      PLAB* plab = ShenandoahThreadLocalData::plab(thread);
+      assert(plab != nullptr, "PLAB should be initialized for %s", thread->name());
+
+      // There are two reasons to retire all plabs between old-gen evacuation passes.
+      //  1. We need to make the plab memory parsable by remembered-set scanning.
+      //  2. We need to establish a trustworthy UpdateWaterMark value within each old-gen heap region
+      ShenandoahGenerationalHeap::heap()->retire_plab(plab, thread);
+      if (_resize && ShenandoahThreadLocalData::plab_size(thread) > 0) {
+        ShenandoahThreadLocalData::set_plab_size(thread, 0);
+      }
+    }
+  }
+};
+
+class ShenandoahGCStatePropagator : public ThreadClosure {
+public:
+  explicit ShenandoahGCStatePropagator(char gc_state) : _gc_state(gc_state) {}
+
+  void do_thread(Thread* thread) override {
+    ShenandoahThreadLocalData::set_gc_state(thread, _gc_state);
+  }
+private:
+  char _gc_state;
+};
+
+class ShenandoahPrepareForUpdateRefs : public HandshakeClosure {
+public:
+  explicit ShenandoahPrepareForUpdateRefs(char gc_state) :
+    HandshakeClosure("Shenandoah Prepare for Update Refs"),
+    _retire(ResizeTLAB), _propagator(gc_state) {}
+
+  void do_thread(Thread* thread) override {
+    _propagator.do_thread(thread);
+    if (ShenandoahThreadLocalData::gclab(thread) != nullptr) {
+      _retire.do_thread(thread);
+    }
+  }
+private:
+  ShenandoahRetireGCLABClosure _retire;
+  ShenandoahGCStatePropagator _propagator;
+};
+
 void ShenandoahHeap::evacuate_collection_set(bool concurrent) {
   ShenandoahEvacuationTask task(this, _collection_set, concurrent);
   workers()->run_task(&task);
 }
 
+void ShenandoahHeap::concurrent_prepare_for_update_refs() {
+  // It's possible that evacuation succeeded, but we could still be cancelled when we get here.
+  // A cancellation at this point means the degenerated cycle must resume from update-refs.
+  set_gc_state_concurrent(EVACUATION, false);
+  set_gc_state_concurrent(WEAK_ROOTS, false);
+  set_gc_state_concurrent(UPDATEREFS, true);
+
+  // This will propagate the gc state and retire gclabs and plabs for threads that require it.
+  ShenandoahPrepareForUpdateRefs prepare_for_update_refs(_gc_state.raw_value());
+
+  // The handshake won't touch non-java threads, so do those separately.
+  Threads::non_java_threads_do(&prepare_for_update_refs);
```

Comment:
Which non-Java threads need to prepare for update refs? (i.e. which make use of this state predicate and/or participate in update refs phase.)

Would be good to document that somewhere (or may be repeat it here).

Reply:
I will change the comment here. Specifically, we are most interested in changing the gc state for the worker threads, but also the control thread and vm thread and any other thread that might touch an oop.